lib/blob/blobstore.c

```diff
@@ -5985,6 +5984,99 @@ spdk_blob_reset_used_clusters_cache(struct spdk_blob *blob)
 		spdk_spin_unlock(&blob->bs->used_lock);
 	}
 }
+
+struct spdk_blob_bitmap_ctx {
+	spdk_blob_cluster_bitmap_complete cb_fn;
+	void *cb_arg;
+	struct spdk_bit_array *bitmap;
+	struct spdk_blob *blob;
+};
+
+static void
+spdk_blob_bitmap_unfreeze_cpl(void *cb_arg, int rc)
+{
+	struct spdk_blob_bitmap_ctx *ctx = (struct spdk_blob_bitmap_ctx *)cb_arg;
+
+	if (rc != 0) {
+		SPDK_ERRLOG("Unfreeze failed, rc=%d\n", rc);
+	}
+
+	ctx->blob->locked_operation_in_progress = false;
+	ctx->cb_fn(ctx->cb_arg, rc, ctx->bitmap);
+	spdk_bit_array_free(&ctx->bitmap);
+	free(ctx);
+}
+
+static void
+spdk_blob_create_bitmap(struct spdk_blob_bitmap_ctx *ctx)
+{
+	struct spdk_blob *blob = ctx->blob;
+	uint64_t i;
+
+	ctx->bitmap = spdk_bit_array_create(blob->active.num_clusters);
+	if (ctx->bitmap == NULL) {
+		SPDK_ERRLOG("Failed to allocate bit array memory for blob #%ld", blob->id);
+		return;
+	}
+
+	for (i = 0; i < blob->active.num_clusters; i++) {
+		if (blob->active.clusters[i]) {
+			spdk_bit_array_set(ctx->bitmap, i);
+		}
+	}
+}
+
+static void
+spdk_blob_bitmap_freeze_cpl(void *cb_arg, int rc)
+{
+	struct spdk_blob_bitmap_ctx *ctx = (struct spdk_blob_bitmap_ctx *)cb_arg;
+
+	if (rc != 0) {
+		ctx->blob->locked_operation_in_progress = false;
+		ctx->cb_fn(ctx->cb_arg, rc, NULL);
+		spdk_bit_array_free(&ctx->bitmap);
+		free(ctx);
+		return;
+	}
+
+	spdk_blob_create_bitmap(ctx);
+	blob_unfreeze_io(ctx->blob, spdk_blob_bitmap_unfreeze_cpl, ctx);
+}
+
+int
+spdk_blob_get_cluster_bitmap(struct spdk_blob *blob, spdk_blob_cluster_bitmap_complete cb_fn,
+			     void *cb_arg)
+{
+	struct spdk_blob_bitmap_ctx *ctx = calloc(1, sizeof(*ctx));
+
+	if (blob->locked_operation_in_progress) {
```

Comment:
blob assert can be put at the begining of the function.

Reply:
Done